### DISCUSSION TITLE
Guaranteed tail call

### DISCUSSION THREAD

**Discussion by MrSmith33** [ldc-developers/ldc#4355] [Ideas]
Since clang 13 there is `musttail` attribute for that
https://clang.llvm.org/docs/AttributeReference.html#musttail
```c
int g(int);
int f(int x) {
    __attribute__((musttail)) return g(x);
}
```

In D I tried to get this via inline LLVM IR
```llvm
extern(C) int foo(int a, void* f) {
    return inlineIR!(`
        %t1 = bitcast i8* %1 to i32 (i32, i8*)*
        %t2 = musttail call i32 %t1(i32 %0, i8* %1)
        ret i32 %t2`, int, int, void*)(a, f);
}
```
But the resulting IR doesn't have the `musttail`
```llvm
define i32 @foo(i32 %a_arg, i8* %f_arg) {
  %t1.i = bitcast i8* %f_arg to i32 (i32, i8*)*
  %t2.i = tail call i32 %t1.i(i32 %a_arg, i8* %f_arg)
  ret i32 %t2.i
}
```

Is it difficult to add this to LDC?

**Comment by JohanEngelen** [comment]
Have you tried `import ldc.attributes;  @llvmAttr("musttail") ...` ?

Ah, I now see that you want it on the call site. I don't think we have any method for that at the moment.

**Reply by MrSmith33** [reply]
Having it on the function itself is ok if it will work.

I tried the `@llvmAttr("musttail")` and it didn't work: 
```d
import ldc.attributes;
@llvmAttr("musttail") int g(int);
int f(int x) {
    return g(x);
}
```
```d
int example.f(int):
        push    rbp
        mov     rbp, rsp
        sub     rsp, 16
        mov     dword ptr [rbp - 4], edi
        mov     edi, dword ptr [rbp - 4]
        call    int example.g(int)@PLT
        add     rsp, 16
        pop     rbp
        ret
```
https://godbolt.org/z/1h31bsTGb

**Reply by JohanEngelen** [reply]
Unfortunately the dlang language does not allow putting UDAs on statements, so the return statement cannot be marked with `@musttail`.  To implement support for this on short term, I propose that if `@musttail` is applied to function `f`, that if its return statement is a function call, it will be a tail call (rather than applying the attribute to `g`). We already have our own semantic pass in addition to standard dlang semantic analysis, so there we can also generate an error if `@musttail` is applied and correct conditions are not met.

**Reply by MrSmith33** [reply]
To me `musttail` on a function is the same logic as `inline` on a function. The call to that function is affected, not calls inside the function. Maybe we could have a pragma that affects return statement?

**Reply by MrSmith33** [reply]
However, I see what you mean. It will only have `musttail` semantics if a call is in return statement, so one can have both tail calls and non-tail calls in the same function.

**Reply by MrSmith33** [reply]
I convinced myself that `musttail` affecting the body is ok.

**Reply by MrSmith33** [reply]
I would appreciate any pointers on how can this be implemented. I would like to work on this feature

**Reply by kinke** [reply]
When wanting to replicate clang here, I guess I'd start with extending the `ReturnStatement` AST node with an LDC-specific flag wrt. musttail, and then setting it when encountering the new pragma (see `gen/pragma.cpp`). Then for codegen in `visit(ReturnStatement*)` (`gen/statements.cpp`), checking that flag, and if it's set *and* the return expression is a call, setting the `musttail` IR attribute there.

**Reply by kinke** [reply]
Adding a flag to `CallExp` directly might be simpler.

**Reply by MrSmith33** [reply]
Are we talking about making `pragma(LDC_musttail)` a statement pragma?

Where do I process statement pragma and how do I get the statement after the pragma?

**Reply by MrSmith33** [reply]
Progress so far <https://github.com/ldc-developers/ldc/compare/master...MrSmith33:ldc:musttail>

**Reply by kinke** [reply]
Hmm I don't think there are any statement-level pragmas yet, although supported according to the lang spec. A `PragmaDeclaration` has a `decl` list of `Dsymbol`s it applies to - but statements are no `Dsymbol`s...

**Reply by kinke** [reply]
Ah wait, that form is a `PragmaStatement` apparently. Which has a `_body` as Statement it applies to.

**Reply by MrSmith33** [reply]
So pragma without `;` is essentially an attribute.

There is `ToIRVisitor.visit(PragmaStatement)` that is not implemented. So setting a flag on `CallExp` can be done from there. Not sure how where call expression code is generated. Is this [the place](https://github.com/ldc-developers/ldc/blob/master/gen/tocall.cpp#L919)?

**Reply by JohanEngelen** [reply]
In my opinion, the handling of the pragma statement should be done during semantic analysis, not during IR generation. That is also where errors should be caught (e.g. pragma(musttail) on a function or regular function call), such that it fits in with all normal semantic analysis (__traits(compiles, ...)).

(note we already have this problem with ldc's UDAs...)

**Reply by kinke** [reply]
> Is this [the place](https://github.com/ldc-developers/ldc/blob/master/gen/tocall.cpp#L919)?

The CallExp codegen calls that function: https://github.com/ldc-developers/ldc/blob/0d4d711aed785e40f8d894187f8be0d361b8589a/gen/toir.cpp#L779-L780

You'd need to add a line similar to https://github.com/ldc-developers/ldc/blob/0d4d711aed785e40f8d894187f8be0d361b8589a/ir/irclass.cpp#L639 after https://github.com/ldc-developers/ldc/blob/0d4d711aed785e40f8d894187f8be0d361b8589a/gen/tocall.cpp#L1067 when passing the tail-call bla through, or ideally in that `call()` helper directly, where you'd need to get hold of the `llvm::CallInst*` from the returned `DValue`.

**Comment by MrSmith33** [comment]
Added PR https://github.com/ldc-developers/ldc/pull/4366

**Reply by MrSmith33** [reply]
Problem is that `callOrInvoke` returns `LLCallBasePtr`, which is base to `InvokeInst` and `CallInst`. `setTailCallKind` is only present on `CallInst`.

**Reply by MrSmith33** [reply]
I could propagate `bool isMustTail` into `callOrInvoke`, but what to do with invoke call? Ideally compiler should produce error in that case.

**Reply by deadalnix** [reply]
Invoke calls are required when there is unwinding code required, such as scope exit/failure , dtors, exception handling, etc...

When they are present, then tail call is not possible. On the other hand, if a function might throw but none of the above is present, it's better to generate a good old call instruction and the call can be a tail call.

**Comment by DmitryOlshansky** [comment]
https://www.mattkeeter.com/blog/2026-04-05-tailcall/
